### ISSUE TITLE
Loss scaling

### ISSUE THREAD

**Issue by tvelovraf** [yukara-ikemiya/modified-shortcut-models-pytorch#2]
Hi! First, thank you for that pytorch implemetation of the paper

I have a question about the loss computing aspect.
Flow-matching loss is computed there
https://github.com/yukara-ikemiya/modified-shortcut-models-pytorch/blob/b4a967e0bcddde45ddba4ba8dc51b86d49657a6b/src/shortcut_models/shortcut.py#L98

and it's averaged across `bs-num_self_consistency` samples and other dimentions.

Self-consistency term is computed there in the same way
https://github.com/yukara-ikemiya/modified-shortcut-models-pytorch/blob/b4a967e0bcddde45ddba4ba8dc51b86d49657a6b/src/shortcut_models/shortcut.py#L103

but it is averaged across `num_self_consistency` samples (and other dimentions). And both loss terms are then added together.

So, I think that loss terms are not scaled in the right way to be added in such manner.

For example, `bs - num_self_consistency = 0.25*bs`. Then in your code:

$$
\Large Loss_{fm} = \frac{1}{0.75bs} * \sum_{i=0.25bs+1}^{bs} [mean((pred_i - target_i)^2)]
$$

$$
\Large Loss_{sc} = \frac{1}{0.25bs} * \sum_{i=1}^{0.25bs} [mean((pred_i - target_i)^2)] 
$$

But in such case self-consistency loss would be weighted to have the same impact as flow-matching one even being computed on less number of samples.

Don't we need to to use $\Large \frac{1}{bs}$ weighting term in loss formulas above instead of  $\Large \frac{1}{0.25bs}$ and  $\Large \frac{1}{0.75bs}$? Or maybe it was made intentionally?

**Comment by yukara-ikemiya** [comment]
It seems like your question got cut off..?

**Comment by tvelovraf** [comment]
Sorry, accidentally sent the question before finished it

**Comment by yukara-ikemiya** [comment]
@tvelovraf 
Thank you for the question.
I was actually thinking the same thing while implementating it myself. In the official implementation, the latter loss calculation (`1/bs`) seems to be used, where each term is weighted by the number of sampes. On the other hand, as you pointed out, my implementation uses the former loss calculation.

[Loss calculation of official codes]
https://github.com/kvfrans/shortcut-models/blob/601004348667094e1b71f30942199759412d4432/train.py#L239

However, as stated in the paper, the self-consistency loss is significantly smaller than the flow-matching loss, so the difference in weighting doesn't have much impact. I believe the adjustment of sample number in batch primarily serves to speed up training.

Conversely, it's unlikely that either weighting scheme is truly optimal for the given training data, so in practice, it might be necessasry to introduce **a new loss weighting parameter** and search for the best setting for each dataset.

(From the paper)
![Image](https://github.com/user-attachments/assets/70bb96f5-fee8-4c5f-8629-d3985d477fdf)

**Comment by tvelovraf** [comment]
@yukara-ikemiya
> Conversely, it's unlikely that either weighting scheme is truly optimal for the given training data, so in practice, it might be necessasry to introduce **a new loss weighting parameter** and search for the best setting for each dataset.

Yes, that sounds very reasonable 

Thank you for the asnwer!
